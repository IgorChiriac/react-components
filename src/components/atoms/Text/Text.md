### Summary

1. Whenever you want to render text on any UI at Zopa, use the `<Text />` component 🙏🏻
2. Its size, weight, white-space, semantics and colour can be customised
3. If you need to render long text, render it within `<p />` tag, otherwise use the defautl `<span />` tag 👮🏻‍♂️
4. Don't use `semibold` for now as we don't have clear specs on when to use it over `bold`
5. Use the colours mindfully 🎨

### Examples

- Variations

```jsx
import { Fragment } from 'react';
import { Text } from '@zopauk/react-components';

<Fragment>
  <Text as="p" mb>
    Long paragraph, rendered within a HTML `p` tag: we give our customers a fair deal as standard and our products are
    built so that we win when you win. Managing your money is no sweat. With handy tools to get stuff done and helpful
    people always at the end of a phone. We listen to what you want and change with your needs.
  </Text>
  <Text>Text as HTML `span`</Text>
</Fragment>;
```

- With white-space

```jsx
import { Fragment } from 'react';
import { Text } from '@zopauk/react-components';

<Fragment>
  <Text mb>I have `margin-bottom` below ⤵️</Text>
  <Text>I'm pushed further down 🙄</Text>
</Fragment>;
```

- Sizes

```jsx
import { Fragment } from 'react';
import { Text } from '@zopauk/react-components';

<Fragment>
  <Text size="large" mb>
    Size 1 ( 16px )
  </Text>
  <Text mb>Size 2 ( 14px, default )</Text>
  <Text size="small">Size 3 ( 12px )</Text>
</Fragment>;
```

- Weights

```jsx
import { Fragment } from 'react';
import { Text } from '@zopauk/react-components';

<Fragment>
  <Text mb>Regular weight</Text>
  <Text weight="semibold" mb>
    Semi-bold weight
  </Text>
  <Text weight="bold" mb>
    Bold weight
  </Text>
</Fragment>;
```

- Colors

```jsx { "props": { "style": { "backgroundColor": "rgb(244, 248, 246)", "border": "none" } } }
import { Fragment } from 'react';
import { Text, colors } from '@zopauk/react-components';

<Fragment>
  <Text color={colors.neutral.white} mb>
    White
  </Text>
  <Text color={colors.neutral.medium} mb>
    Medium
  </Text>
  <Text color={colors.neutral.dark} mb>
    Dark
  </Text>
  <Text color={colors.semantic.success} mb>
    Success
  </Text>
  <Text color={colors.semantic.error}>Error</Text>
</Fragment>;
```
