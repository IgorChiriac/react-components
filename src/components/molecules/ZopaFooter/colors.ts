export const neutral900 = '#1C2139';
export const neutral200 = '#A9ACBA';
export const neutral400 = '#6F748B';
export const neutral800 = '#282F52';
