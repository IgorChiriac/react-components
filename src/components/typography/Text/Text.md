Text component

### Text sizes

```jsx
<div>
  <Text size="xl">Extra large</Text>
</div>
<div>
  <Text size="l">Large</Text>
</div>
<div>
  <Text size="m">Medium (default)</Text>
</div>
<div>
  <Text size="s">Small</Text>
</div>
<div>
  <Text size="xs">Extra small</Text>
</div>
```

### Text font weights

```jsx
<div>
  <Text fw="normal">Normal (default)</Text>
</div>
<div>
  <Text fw="bold">Bold</Text>
</div>
```

### Text custom colors

```jsx
<div>
  <Text>Default</Text>
</div>
<div>
  <Text color="#00B9A7">Teal</Text>
</div>
<div>
  <Text color="#5E35B1">Indigo</Text>
</div>
```

### Text as different html elements

```jsx
<div>
  <Text as="span">Text as span tag (default)</Text>
</div>
<div>
  <Text as="p">Text as p tag</Text>
</div>
```
