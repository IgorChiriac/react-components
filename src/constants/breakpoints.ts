export const breakpoints = {
  desktop: 1280,
  phone: 600,
  tablet: 720,
} as const;

export type TDeviceSizes = keyof typeof breakpoints;
