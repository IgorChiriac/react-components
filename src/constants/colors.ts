export type TTextHexColors = '#FFFFFF' | '#C5C6CD' | '#0D0A38' | '#1EC06A' | '#EE0505';
export type THeadingHexColors = '#FFFFFF' | '#0D0A38';
export type TLinkHexColors = '#FFFFFF' | '#0D0A38';

export const colors = {
  base: {
    primary: '#00B9A7',
    secondary: '#4A3EDE',
  },
  neutral: {
    white: '#FFFFFF',
    light: '#EAEAEE',
    medium: '#C5C6CD',
    dark: '#0D0A38',
  },
  semantic: {
    success: '#1EC06A',
    alert: '#FFB428',
    error: '#EE0505',
  },
};
