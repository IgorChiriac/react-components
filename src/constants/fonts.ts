export const openSans: string =
  '"Open Sans", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"';
export const alverata: string = 'Alverata, "Arial Black", Verdana, sans-serif';
