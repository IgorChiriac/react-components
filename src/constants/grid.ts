const grid = {
  breakpoints: {
    l: 992,
    m: 768,
    s: 576,
    xl: 1200,
    xs: 0,
  },
  cols: 12,
  gutter: 16,
  width: {
    l: 960,
    m: 720,
    s: 540,
    xl: 1140,
  },
};

export default grid;
