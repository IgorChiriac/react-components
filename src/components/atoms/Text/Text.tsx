import React from 'react';
import { HTMLAttributes } from 'react';
import styled from 'styled-components';
import { typography } from '../../../constants/typography';

export interface ITextProps extends HTMLAttributes<HTMLSpanElement | HTMLParagraphElement> {
  /**
   * The weight of the rendered text. Avoid the use of `semibold` as for now.
   * @default 400
   */
  weight?: keyof typeof typography.weights;
  /**
   * The size you want to render your text at, currently only `12px` | `14px` | `16px` supported.
   * @default 'medium'
   */
  size?: keyof typeof typography.sizes.text;
  /**
   * The HTML5 tag you want to render your text on, currently only `<span>` and `<p>` are supported.
   * @default span
   */
  as?: 'span' | 'p' | 'label';
  /**
   * Whether to add some margin below the rendered text or not. Use it to give meaningful white-space.
   * @default false
   */
  mb?: boolean;
}

const Text = styled.span<ITextProps>`
  margin: 0;
  letter-spacing: 0;

  ${({ mb = false }) =>
    mb &&
    `
    display: block;
    margin-bottom:   24px
  `};

  line-height: ${typography.lineHeight};
  font-family: ${typography.primary};
  font-weight: ${({ weight = 'regular' }) => typography.weights[weight]};
  font-size: ${({ size = 'medium' }) => {
    return typography.sizes.text[size];
  }};
`;

const TextWrap: React.FunctionComponent<ITextProps> = props => <Text {...props} />;

export default TextWrap;
