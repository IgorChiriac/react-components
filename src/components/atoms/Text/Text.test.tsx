import { axe } from 'jest-axe';
import React from 'react';
import { render } from '@testing-library/react';
import Text from './Text';
import { colors } from '../../../constants/colors';

describe('<Text />', () => {
  it('renders without  a11y violations', async () => {
    const { container } = render(<Text>Text</Text>);
    const results = await axe(container.innerHTML);

    expect(results).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });

  it('can render with bottom white-space', () => {
    const { container } = render(
      <Text as="p" mb>
        Long paragraph with some bottom white-space to prevent text next to it colliding.
      </Text>,
    );

    expect(container.firstChild).toHaveStyleRule('margin', '0');
  });

  it.each`
    tag
    ${'span'}
    ${'p'}
  `('it can render as an HTML $tag tag', ({ tag }) => {
    const { container } = render(<Text as={tag}>Text</Text>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it.each`
    size       | pixels
    ${'lead'}  | ${'20px'}
    ${'base'}  | ${'16px'}
    ${'small'} | ${'14px'}
  `('can render at different sizes:  $size – $pixels', ({ size, pixels }) => {
    const { container } = render(<Text size={size}>Text</Text>);
    expect(container.firstChild).toHaveStyleRule('font-size', pixels);
  });

  it.each`
    name          | weight
    ${'regular'}  | ${'400'}
    ${'semibold'} | ${'600'}
    ${'bold'}     | ${'700'}
  `('can render at different weights:  $name', ({ name, weight }) => {
    const { container } = render(<Text weight={name}>Text</Text>);
    expect(container.firstChild).toHaveStyleRule('font-weight', weight);
  });

  it.each`
    hex                   | name
    ${colors.white}       | ${'White'}
    ${colors.grey}        | ${'Grey'}
    ${colors.greyDarkest} | ${'Grey Darkest'}
    ${colors.success}     | ${'Success'}
    ${colors.alert}       | ${'Alert'}
  `('can render in different colors: $name – $hex', ({ hex }) => {
    const { container } = render(<Text color={hex}>Text</Text>);
    expect(container.firstChild).toHaveStyleRule('color', hex);
  });
});
