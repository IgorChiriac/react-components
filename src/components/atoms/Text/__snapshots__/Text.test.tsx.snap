// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> can render at different sizes:  "large" – "16px" 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  font-size: 16px;
}

<span
  class="c0"
>
  Text
</span>
`;

exports[`<Text /> can render at different sizes:  "medium" – "14px (default)" 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  font-size: 14px;
}

<span
  class="c0"
>
  Text
</span>
`;

exports[`<Text /> can render at different sizes:  "small" – "12px" 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  font-size: 12px;
}

<span
  class="c0"
>
  Text
</span>
`;

exports[`<Text /> can render at different weights:  "bold" 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 700;
  font-size: 14px;
}

<span
  class="c0"
>
  Text
</span>
`;

exports[`<Text /> can render at different weights:  "regular" 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  font-size: 14px;
}

<span
  class="c0"
>
  Text
</span>
`;

exports[`<Text /> can render at different weights:  "semibold" 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 600;
  font-size: 14px;
}

<span
  class="c0"
>
  Text
</span>
`;

exports[`<Text /> it can render as an HTML "p" tag 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  font-size: 14px;
}

<p
  class="c0"
>
  Text
</p>
`;

exports[`<Text /> it can render as an HTML "span" tag 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  font-size: 14px;
}

<span
  class="c0"
>
  Text
</span>
`;

exports[`<Text /> renders without  a11y violations 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  font-size: 14px;
}

<div>
  <span
    class="c0"
  >
    Text
  </span>
</div>
`;
